### PR TITLE
Pre-verify company email address

### PR DESCRIPTION
**Pull request by richtier**
[this task](https://uktrade.atlassian.net/browse/ED-1706)

### CODE REVIEW

File: company/migrations/0041_add_search_index.py [hunk 1]
```diff
@@ -24,16 +24,17 @@ def remove_company_search_index(apps, schema_editor):
...
     ]
 
     operations = [
+        # calling rebuild elasticserach index uses the Company model. Once new
+        # fields are added to the Company model, but the schema has not yet
+        # been migrated then the "rebuild elasticserach index" will fail
         migrations.RunPython(
-            add_company_search_index_and_populate,
-            remove_company_search_index
+            migrations.RunPython.noop, migrations.RunPython.noop
```
Comment (konradko) [inline company/migrations/0041_add_search_index.py]: let's create a helper method that will take a historical version of the model to populate ES, using `apps.get_model`. Created https://uktrade.atlassian.net/browse/ED-1708 for it

File: company/signals.py [hunk 2]
```diff
@@ -5,12 +5,14 @@
...
-    has_invalid_address = not instance.has_valid_address()
-    is_already_sent = instance.is_verification_letter_sent
-    if is_disabled or is_already_sent or has_invalid_address:
+    skip_send_if_any_true = [
+        not settings.FEATURE_VERIFICATION_LETTERS_ENABLED,
+        instance.is_verification_letter_sent,
+        instance.verified_with_trade_association,
+        not instance.has_valid_address(),
+    ]
+    if any(skip_send_if_any_true):
```
Comment (konradko) [inline company/signals.py]: `any(skip_send_if_any_true)` is a bit mind-bending, maybe:
```
skip_send = any(
     not settings.FEATURE_VERIFICATION_LETTERS_ENABLED,
     instance.is_verification_letter_sent,
     instance.verified_with_trade_association,
     not instance.has_valid_address(),
)
if skip_send:
```

File: enrolment/models.py [hunk 3]
```diff
@@ -1,22 +1,16 @@
-from django.conf import settings
-from django.core.signing import Signer
 from django.db import models
 from directory_validators import enrolment as shared_validators
 
 from api.model_utils import TimeStampedModel
 
 
-class TrustedSourceSignupCode(TimeStampedModel):
+class PreVerifiedCompany(TimeStampedModel):
```
Comment (konradko) [inline enrolment/models.py]: Please rename to `PreVerifiedSignup` or `PreVerifiedEnrolment` as discussed (*Company sounds like a different kind of a company).

File: enrolment/templates/admin/enrolment/company_csv_upload_form.html [hunk 4]
```diff
@@ -5,7 +5,7 @@
 <div class="breadcrumbs">
 <a href="{% url 'admin:index' %}">{% trans 'Home' %}</a>
 &rsaquo; <a href="{% url 'admin:app_list' app_label='enrolment' %}">Trusted source signup codes</a>
-&rsaquo; <a href="{% url 'admin:generate_trusted_source_upload' %}">Generate source signup codes</a>
+&rsaquo; <a href="{% url 'admin:pre-verify-companies' %}">Generate source signup codes</a>
```
Comment (konradko) [inline enrolment/templates/admin/enrolment/company_csv_upload_form.html]: `Add pre-verified signups`?

File: company/models.py [hunk 5]
```diff
@@ -76,6 +76,7 @@ class Company(TimeStampedModel):
         blank=True,
         default=helpers.generate_verification_code,
     )
+    verified_with_trade_association = models.BooleanField(default=False)
```
Comment (konradko) [inline company/models.py]: this functionality is not specific to trade associations - we should use a more generic name, maybe `verified_with_trusted_source`, though that sounds vague. Something pointing at `PreVerifiedEnrolment` usage would leave a good trace, so maybe `verified_with_preverified_enrolment` would oddly be best?

File: company/signals.py [hunk 6]
```diff
@@ -5,13 +5,14 @@
...
 
 def send_first_verification_letter(sender, instance, *args, **kwargs):
-    is_disabled = not settings.FEATURE_VERIFICATION_LETTERS_ENABLED
-    has_invalid_address = not instance.has_valid_address()
-    is_already_sent = instance.is_verification_letter_sent
-    if is_disabled or is_already_sent or has_invalid_address:
-        return
-
-    send_verification_letter(company=instance)
+    should_send_letter = all([
```
Comment (konradko) [inline company/signals.py]: 👍

File: enrolment/tests/test_views.py [hunk 7]
```diff
@@ -103,58 +90,55 @@ def test_enrolment_create_disables_signup_code_single_code():
...
+    response = api_client.post(url, VALID_REQUEST_DATA, format='json')
 
-    trusted_source_signup_code_one.refresh_from_db()
-    trusted_source_signup_code_two.refresh_from_db()
-    assert trusted_source_signup_code_one.is_active is False
-    assert trusted_source_signup_code_two.is_active is False
+    assert response.status_code == status.HTTP_201_CREATED
+
+    company = Company.objects.last()
+    trusted_source_signup_code.refresh_from_db()
```
Comment (konradko) [inline enrolment/tests/test_views.py]: references old name

File: enrolment/tests/test_views.py [hunk 8]
```diff
@@ -103,58 +90,55 @@ def test_enrolment_create_disables_signup_code_single_code():
...
+    assert trusted_source_signup_code_two.is_active is False
+    assert company.verified_with_trade_association is True
 
 
 @pytest.mark.django_db
 @patch('api.signature.SignatureCheckPermission.has_permission', Mock)
-def test_enrolment_create_disables_signup_code_multiple_codes():
-    trusted_source_signup_code_one = TrustedSourceSignupCodeFactory.create(
-        company_number=VALID_REQUEST_DATA['company_number']
+def test_enrolment_create_signup_code_single_code_different_email():
```
Comment (konradko) [inline enrolment/tests/test_views.py]: still mentions`code`

File: enrolment/tests/test_views.py [hunk 9]
```diff
@@ -103,58 +90,55 @@ def test_enrolment_create_disables_signup_code_single_code():
...
-        is_active=False
+def test_enrolment_create_disables_signup_code_multiple_codes():
+    trusted_source_signup_code_one = PreVerifiedEnrolmentFactory.create(
+        company_number=VALID_REQUEST_DATA['company_number'],
+        email_address=VALID_REQUEST_DATA['contact_email_address'],
     )
-
-    url = reverse(
-        'trusted-source-signup-code', kwargs={'code': trusted_source_code.code}
+    trusted_source_signup_code_two = PreVerifiedEnrolmentFactory.create(
```
Comment (konradko) [inline enrolment/tests/test_views.py]: no codes now, needs a new name

File: enrolment/templates/admin/enrolment/change_list.html [hunk 10]
```diff
@@ -12,8 +12,8 @@
           </a>
         </li>
         <li>
-          <a href="{% url 'admin:generate_trusted_source_upload' %}">
-            Generate signup links
+          <a href="{% url 'admin:pre-verify-companies' %}">
+            Upload emails form trusted source
```
Comment (konradko) [inline enrolment/templates/admin/enrolment/change_list.html]: typo?

File: api/urls.py [hunk 11]
```diff
@@ -55,11 +53,6 @@
         name='enrolment'
     ),
     url(
-        r'trusted-code/(?P<code>.*)/$',
```
Comment (konradko) [inline api/urls.py]: we still need an endpoint for checking if the company-email pair was pre-verified

File: company/models.py (hunk 5 above)
Comment (richtier) [inline company/models.py]: `verified_with_preverified_enrolment` is good. It hints that the field is functionally related to `verified_with_code` (both are "is the company somehow verified" fields), and hints about the model. Will change.

Comment (konradko) [inline company/models.py]: Maybe we should FK to `PreVerifiedEnrolment` too?

File: enrolment/views.py [hunk 12]
```diff
@@ -27,16 +30,23 @@ def post(self, request, *args, **kwargs):
...
+            email_address=self.request.GET.get('email_address'),
+            company_number=self.request.GET.get('company_number'),
+        )
+        # an email+number pair can be pre-verified multiple times because e.g.,
+        # multiple trade organisations vouched for the same company. If there
+        # are multiple, we just need the first one.
+        preverified_enrolment = queryset.first()
+        if not preverified_enrolment:
+            raise Http404()
+        return preverified_enrolment
```
Comment (richtier) [inline enrolment/views.py]: an alternative to this is to have this view be a list view that lists all the `preverified_enrolment` objects for the specified company number and email address - but I think what we have here is fine.

File: api/urls.py (hunk 11 above)
Comment (richtier) [inline api/urls.py]: I've re-added the view but renamed the url to reflect the change in model name,

File: enrolment/views.py (hunk 12 above)
Comment (konradko) [inline enrolment/views.py]: There should be only one active `preverified_enrolment` allowed for given company number and email pair. We should add a DB constraint for that

**Pull request merged**